### PR TITLE
Use party's quest key in cron

### LINKED ISSUE #4174: Party member did not damage boss
Hello,

One of the members of my party should have done a lot of damage to the boss that we are fighting, but no damage was done. My id is eefad08c-4193-42dd-bea8-46b81837cccc, the group id is  d5913e14-ecb9-4cd4-a56a-b5e74a03f7e7, and the party member Natrin whose attack did not happen is d3d82450-a2a8-4040-90c3-d791b5481390.

Thanks!
-Phil

image

---

_edit by admin:_
# What to do if you have this bug:

**If you think you should have damaged your quest boss or should have found items for a collection quest, but you don't see any messages in party chat about your quest progress:**
1. Be aware that quest progress is not applied until the day **after** you have ticked off some To-Dos and Dailies. So if the quest started today, you will see quest progress tomorrow, and this is not a bug. You do not need to report anything here.
2. If your quest started yesterday or earlier but you don't see messages in party chat about your quest progress, click the "Fetch Recent Messages" button. If you now see a message about your own quest progress, then everything is working correctly and you can stop reading here.
3. If you still don't see your quest progress message, look at the quest box on the party page. Does your name appear in the list of participants? If it does not, then you are not participating in the quest. This could be because you joined the party after the quest started OR because the quest owner started the quest before you accepted the invitation. If the quest has not progressed very far, the quest owner might want to abort the quest and then restart it (the quest scroll will reappear in their Inventory for re-use). You do **not** need to report this situation here and there is nothing that admins can do to help.
4. If none of those points are applicable to you, please **_add a comment to this issue with your unique user ID (UUID)**_, which you can find on your [Settings -> API](https://habitrpg.com/#/options/settings/api) page. If others of your party members have the same problem, please tell us that too, but you don't need to give us their UUIDs (we can find them in your quest data in the database). We will then be able to fix the problem for you.

If you aren't sure about any of this, ask a question in the Newbies Guild ([Help -> Ask a Question](https://habitrpg.com/#/options/groups/guilds/5481ccf3-5d2d-48a9-a871-70a7380cee5a)) and we'll give you advice there.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/5428175-party-member-did-not-damage-boss?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F68393&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #7567: Cron should pull quest key from party
When cron runs, it pulls the quest key from the user. The key can get out of sync with the group for unknown reasons. Cron already loads the party so it can use the key from the party.

### PR DESCRIPTION
**Pull request by crookedneighbor**
Closes #7567
Closes #4174
- [x] Write tests for Group model quest methods
- [x] Run collection quest calculations at cron [see this comment](https://github.com/HabitRPG/habitrpg/issues/7567#issuecomment-222881411)
- [x] Write migration for users currently on a collection quest
- [x] Test behavior with transitioning from old quest data to new
- [x] Update DDT for new `quest.collect` data type
- [x] Revert temporary code when migration is done

### CODE REVIEW

File: common/script/fns/randomDrop.js [hunk 1]
```diff
@@ -38,15 +36,12 @@ module.exports = function randomDrop (user, options, req = {}) {
...
-    dropK = randomVal(user, quest.collect, {
-      key: true,
-    });
-    if (!user.party.quest.progress.collect[dropK])
-      user.party.quest.progress.collect[dropK] = 0;
-    user.party.quest.progress.collect[dropK]++;
+  if (predictableRandom(user, user.stats.gp) < chance) {
+    // TEMPORARY, remove once collection migration completes
+    if (typeof user.party.progress.collect !== 'number') {
+      user.party.quest.progress.collecti = 0;
```
Comment (paglias) [inline common/script/fns/randomDrop.js]: typo?

Comment (crookedneighbor) [inline common/script/fns/randomDrop.js]: Yep, definitely. As was `user.party.progress`

**Comment by crookedneighbor** [comment]
@paglias I _think_ this is ready to go. I added some safety precautions to the code that can be removed once the migration is complete: https://github.com/HabitRPG/habitrpg/pull/7570/commits/b3b437e593b5e0d810bc3a7afa327b40d21c4429

@Alys What do you need from me for the DDT changes?

**Comment by Alys** [comment]
@crookedneighbor Just reminder that I have to do stuff, which you've now done. :)

If possible, it would be good if this could be deployed at a time when I'm available to push up the new version of the DDT. I'll be available for another 7 hours today, then unavailable for 9 hours, then available for maybe 2 hours. After that is okay too, but I've run out of fingers to count on. :)

**Comment by Alys** [comment]
Actually, don't worry about coordinating the deployment with me, as long as you don't deploy within the next couple of hours. I'm adjusting the DDT so it can handle either collection data format.

File: website/server/models/user.js [hunk 2]
```diff
@@ -376,7 +376,8 @@ export let schema = new Schema({
       progress: {
         up: {type: Number, default: 0},
         down: {type: Number, default: 0},
-        collect: {type: Number, default: 0},
+        // TEMPORARY - Switch type to Number after migration
+        collect: {type: Schema.Types.Mixed, default: 0},
```
Comment (paglias) [inline website/server/models/user.js:380]: 👍

**Comment by paglias** [comment]
@crookedneighbor go ahead!

**Pull request merged**